### ISSUE TITLE
Cancelling a jet job's future does not cancel the job

### ISSUE THREAD

**Issue by wfhartford** [hazelcast/hazelcast#23977]
The jet `Job` interface has a `getFuture()` method which returns a `CompletableFuture`. One can wait on this future for the job to complete, however, the future's `cancel()` method throws an `UnsupportedOperationException`. The `Job.getFuture()` method's documentation indicates that the future is not cancellable, so it is reasonable to say that this is not a bug:
> Gets the future associated with the job. The returned future is not cancellable. To cancel the job, the cancel() method should be used.

However, I personally consider this a bug since it breaks the contract of the `Future.cancel()` method. The `Job` associated with the `Future` is cancellable (by the `Job.cancel()` method), so in my opinion, the `Future.cancel()` method should cancel it's associated `Job`.

Is there a compelling reason that the `CompletableFuture.cancel()` method cannot cancel the jet `Job`?

This deviation from the `Future.cancel()` contract is especially acute when using Hazelcast and Jet from Kotlin where cancellation is expected to work in a certain way to allow for structured concurrency. Kotlin coroutines provides an extension function `CompletionStage.await()` which suspends until the `CompletionStage` finishes. Kotlin's structured concurrency semantics mean that if a coroutine is cancelled, all underlying jobs are expected to be cancelled. `CompletionStage.await()` achieves this by calling the `CompletableFuture.cancel()` method.

**Expected behavior**
A call to `Job.getFuture().cancel()` should cancel the job.

**To Reproduce**
In Java, the following code will throw an `UnsupportedOperationException` and leave the job running:
```java
Job job = hazelcast.getJet().newJob(pipeline)
job.getFuture().cancel()
```
In Kotlin, the following code will _log_ an `UnsupportedOperationException` and leave the job running:
```kotlin
  coroutineScope {
    val job = hz.jet.newJob(pipe)
    val deferred = async { job.future.await() }
    deferred.cancel()
  }
```

In either of these situations, a reasonable developer who understands working with Java or Kotlin concurrency mechanisms, would expect the job to be cancelled. 

**Additional context**
I understand that the future's inability to cancel the job is documented in the `Job.getFuture()` method JavaDoc, and a reasonable response to this bug report might be "RTFM", however, developers have expectations when dealing with concepts they are familiar with, such as Java's `Future` and related classes, and Kotlin's structured concurrency. It might be reasonable to consider this a feature request rather than a bug report, regardless, I think that it's important for software to behave the way reasonable developers expect it to.

I would be happy to look into submitting a pull request to address this if it were likely to be accepted.

**Workaround**
I've implemented the following Kotlin function which suspends waiting for the job to complete and implements the expected cancellation behavior:
```kotlin
suspend fun Job.await() {
  suspendCancellableCoroutine { cont: CancellableContinuation<Unit> ->
    future.whenComplete { _, t ->
      if (t == null) cont.resume(Unit)
      else cont.resumeWithException(t)
    }
    cont.invokeOnCancellation { cancel() }
  }
}
```

**Comment by viliam-durina** [comment]
If I remember correctly, it is because we implemented the cancellation differently than `CompletableFuture`. If you cancel a standard future, it's immediately cancelled and its `join` method throws `CancellationException` (unless the future was already done, of course). But if you read [our `cancel()` documentation](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/jet/Job.java#L104-L108), even if you cancel, it might continue to run. Of course, this can happen only under failure scenarios going on in the cluster. You can think of it as of a "distributed cancellation", and it has more challenges than a local cancellation.

In our interpretation, `Job.cancel()` just initiates the cancellation, and returns. You can then `join` the job and wait until it actually terminates. I think with this inconsistency we're paying the price for reusing Java APIs and introducing subtle differences. Perhaps we should have never had the `getFuture()` method, but somebody thought it might be useful to provide some standard API, but at least explicitly broke the `cancel` method, which behaves very differently. We're having subtle differences also in our `java.util.Map` implementation, the `IMap`.

I'm not exactly sure what your workaround does (I'm not a Kotlin dev), but I guess it provides a future that simply calls `Job.cancel()` when cancelled. It should actually call `cancel()`, then monitor the job status, and if it didn't terminate during some time, retry, until it's really terminated. See also [this method](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/jet/core/JetTestSupport.java#L404-L433) which we're using in tests.
